core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java

```diff
@@ -596,19 +654,37 @@ public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         EnvelopePart that = (EnvelopePart) o;
-        return (meta == that.meta
+        return (attrs.equals(that.attrs)
                 && Arrays.equals(positions, that.positions)
                 && Arrays.equals(speeds, that.speeds)
                 && Arrays.equals(timeDeltas, that.timeDeltas));
     }
 
     @Override
     public int hashCode() {
-        int result = System.identityHashCode(meta);
+        int result = attrs.hashCode();
         result = 31 * result + Arrays.hashCode(positions);
         result = 31 * result + Arrays.hashCode(speeds);
         result = 31 * result + Arrays.hashCode(timeDeltas);
         return result;
     }
+
+    @Override
+    public String toString() {
+        var builder = new StringBuilder("EnvelopePart { ");
+        boolean isFirst = true;
+        for (var item : attrs.entrySet()) {
+            if (isFirst)
+                isFirst = false;
+            else
+                builder.append(", ");
```

Comment:
Minor nitpick, but there are easier ways to print delimited items such as [`String.join`](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.CharSequence...-) or [`StringJoiner`](https://docs.oracle.com/javase/8/docs/api/java/util/StringJoiner.html)

A stream -> map -> `Collectors.joining` could also be fitting here.